### PR TITLE
Improve handling tangents when exporting code

### PR DESCRIPTION
**Pull request by kevinwuTT**
### Ticket
https://github.com/tenstorrent/pytorch2.0_ttnn/issues/1065

### Problem description
Gaussian splatting training exposed an issue with handling tangent nodes when exporting code for profiling. 

Based on my observation, tangents are paired with primals. Primals are the inputs while tangents are information about the "direction" of the primals. Which means that if there are no primals in the output list, there are also no tangents.  (More information here: https://docs.pytorch.org/tutorials/intermediate/forward_ad_usage.html) In the lowered aten graphs, the list of tangents always precedes the list of primals. Primals are always named with the text "primals_{index}". However, tangents are only named "tangents_{index}" in the parameter list for subsequent functions, and not named this way in the output list. However, the order of these tangents and primals are always consistent. 

In other words: 

```
[tangents], [primals], [other outputs] = forward_1(...)
```
Actual naming example:
```
stack, add_24, clamp, select, primals_3, primals_4, primals_9, ... = forward(...)
```
Next forward call:
```
[outputs] = forward_2(..., [tangents])
```
Actual naming example:
```
... = forward(..., tangents_1, tangents_2, tangents_3, tangents_4)
```
You can map:
```
tangents_1 = stack
tangents_2 = add_24
tangents_3 = clamp
tangents_4 = select
```
***
However, in another example:
```
mul_1, ge, sigmoid, exp, div, cat, mm_2, primals_4, primals_7, primals_8, ... = forward(...)

... = forward(..., tangents_1, tangents_2, tangents_3, tangents_4, tangents_5, tangents_6)
```
The number of nodes before the first primal in the first forward function is 7, but there are only 6 tangents passed into the second forward function.

During compilation, Pytorch generates a mask that labels which of the output nodes before the first primal node is a tangent: [traced_function_transforms.py](https://github.com/pytorch/pytorch/blob/30587195d314eb5eb02ce63f39a9be4c943629ef/torch/_functorch/_aot_autograd/traced_function_transforms.py#L159-L174). This implementation is internal to Pytorch, so I extract a snippet to use in our backend. In the above example, the mask would look like: 
```
[True, False, True, True, True, True, True]
```
This corresponds to a mapping of:
```
tangents_1 = mul_1
tangents_2 = sigmoid
tangents_3 = exp
tangents_4 = div
tangents_5 = cat
tangents_6 = mm_2
```
with `ge` being dropped.
***

Additional notes:
* Not all outputs from one forward function are passed to the immediate next forward function. Sometimes it could be several functions after. Therefore, it is important to give them unique names to prevent overwrite.
* Tangents and other nodes from the same output list should go to the same forward function.

### What's changed
* Add a preprocessing step for the aten graph to attach additional metadata for tangent nodes
* Use the new metadata to modify the naming of input and output variables so they are passed correctly across various forward function calls

### TODO:
- [ ] This PR currently does not contain the gaussian-splatting code. It may be better to have a separate PR for that.

### CODE REVIEW

File: torch_ttnn/handle_tangents.py
[the remark this answers is not in the document]
Comment (kevinwuTT) [reply torch_ttnn/handle_tangents.py:1]: > Not sure I entirely understand when tangents are args to forward functions

My understanding is that tangents are generated when using "forward-mode" auto differentiation. Whereas, for "reverse-mode" auto differentiation, there is also a backward function but doesn't compute tangents. So for this specific setup of this model, it seems like forward-mode is used. 

> do we need to update the analysis that marks arguments as either primals, arguments, or buffers to also include tangents? (And also update passes that use that information to split inputs for data parallel, for example)

I don't think we need to for now since Pytorch runtime automatically passes the tangents to the proper sub-graphs. 

> Would this be better in the analysis directory or stored where other transforms are?

There's an analysis directory under passes, but this isn't exactly a Pass. But I do agree we should maybe have a separate subdirectory. `handle_input_aliasing.py` and `handle_tangents.py` are both pre-processing modules, so maybe a new directory called `preprocessing` directly under `torch_ttnn`?

File: torch_ttnn/handle_tangents.py [hunk 1]
```diff
@@ -23,13 +23,12 @@ def mark_output_as_tangents(graph: torch.fx.Graph) -> torch.fx.Graph:
     # tangents are nodes before the first primal.
     # If primals do not exist, then nothing will change. We can prepend nodes in advance.
```
Comment (jmalone-tt) [inline torch_ttnn/handle_tangents.py:24]: Would be good to add a TODO about marking tangents in the input_analysis_pass if we want to support multi device for training at some point

**Review by jmalone-tt — COMMENTED** [review]
Added some more questions after our discussion today. Once these are fixed, I think it will be good to merge!
